{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/hail/context.py:350: UserWarning:\n",
      "\n",
      "Using hl.init with a default_reference argument is deprecated. To set a default reference genome after initializing hail, call `hl.default_reference` with an argument to set the default reference genome.\n",
      "\n",
      "/opt/conda/lib/python3.10/site-packages/hailtop/aiocloud/aiogoogle/user_config.py:43: UserWarning:\n",
      "\n",
      "Reading spark-defaults.conf to determine GCS requester pays configuration. This is deprecated. Please use `hailctl config set gcs_requester_pays/project` and `hailctl config set gcs_requester_pays/buckets`.\n",
      "\n",
      "Running on Apache Spark version 3.5.3\n",
      "SparkUI available at http://all-of-us-27410-m.us-central1-a.c.terra-vpc-sc-f223a39d.internal:35713\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.134-952ae203dbbe\n",
      "LOGGING: writing to /home/jupyter/workspaces/asdproject/hail-20251207-2250-0.2.134-952ae203dbbe.log\n",
      "2025-12-07 23:31:41.130 Hail: INFO: merging 13445 files totalling 300.4M...\n",
      "2025-12-07 23:31:58.091 Hail: INFO: while writing:\n",
      "    gs://fc-secure-b8cbf1c7-c51b-44be-b4f7-4f88082bda65/data/aou_sfari_variants_fast.csv\n",
      "  merge time: 16.959s\n",
      "2025-12-07 23:32:14.495 Hail: WARN: entries(): Resulting entries table is sorted by '(row_key, col_key)'.\n",
      "    To preserve row-major matrix table order, first unkey columns with 'key_cols_by()'\n",
      "2025-12-08 00:06:22.013 Hail: INFO: merging 13445 files totalling 3.0G...\n",
      "2025-12-08 00:07:06.960 Hail: INFO: while writing:\n",
      "    gs://fc-secure-b8cbf1c7-c51b-44be-b4f7-4f88082bda65/data/aou_sfari_genotypes_long.csv\n",
      "  merge time: 44.947s\n"
     ]
    }
   ],
   "source": [
    "# 1. INITIALIZE HAIL (Standard AoU Setup)\n",
    "hl.init(default_reference='GRCh38')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Gene Map to filter intervals...\n"
     ]
    }
   ],
   "source": [
    "bucket = os.getenv(\"WORKSPACE_BUCKET\")\n",
    "wgs_path = os.getenv(\"WGS_EXOME_SPLIT_HAIL_PATH\") \n",
    "\n",
    "asd_gene_target_list = {\n",
    "    \"SHANK3\", \"SCN2A\", \"MECP2\", \"CHD8\", \"SYNGAP1\", \"NRXN1\", \"PTEN\", \"SCN1A\", \"CNTNAP2\", \"ARID1B\", \n",
    "    \"ADNP\", \"DYRK1A\", \"GRIN2B\", \"CHD2\", \"FOXP1\", \"ANKRD11\", \"STXBP1\", \"AUTS2\", \"TCF4\", \"FMR1\", \n",
    "    \"GRIN2A\", \"POGZ\", \"PCDH19\", \"KCNQ2\", \"SCN8A\", \"TSC2\", \"CACNA1C\", \"CDKL5\", \"FOXG1\", \"RELN\", \n",
    "    \"DDX3X\", \"SLC6A1\", \"ASXL3\", \"SHANK2\", \"FOXP2\", \"IQSEC2\", \"MEF2C\", \"MED13L\", \"NLGN3\", \"KMT2A\", \n",
    "    \"TRIO\", \"DMD\", \"UBE3A\", \"CHD7\", \"SATB2\", \"GABRB3\", \"KDM5C\", \"OXTR\", \"SETD5\", \"CACNA1A\", \n",
    "    \"MBD5\", \"TBR1\", \"NF1\", \"ASH1L\", \"CTNNB1\", \"EHMT1\", \"KMT2C\", \"MYT1L\", \"RBFOX1\", \"NLGN4X\", \n",
    "    \"TRIP12\", \"CREBBP\", \"MET\", \"ANK2\", \"CUL3\", \"HNRNPU\", \"PPP2R5D\", \"DYNC1H1\", \"NEXMIF\", \"NSD1\", \n",
    "    \"RAI1\", \"ATRX\", \"SETBP1\", \"TSC1\", \"DEAF1\", \"DISC1\", \"ANK3\", \"CACNA1H\", \"DNMT3A\", \"KAT6A\", \n",
    "    \"MTOR\", \"TCF20\", \"TTN\", \"AHDC1\", \"BCL11A\", \"CACNA1D\", \"GRIN1\", \"KCNMA1\", \"TBL1XR1\", \"WAC\", \n",
    "    \"KCNB1\", \"PTPN11\", \"VPS13B\", \"IL1RAPL1\", \"KMT5B\", \"NAA15\", \"ZBTB20\", \"CACNA1G\", \"CASK\", \"CNTN4\", \n",
    "    \"EP300\", \"NR2F1\", \"NRXN3\", \"SETD2\", \"SLC6A4\", \"SMARCA2\", \"HUWE1\", \"MAGEL2\", \"PHIP\", \"SYN1\", \n",
    "    \"CTCF\", \"DPP6\", \"KDM5B\", \"NBEA\", \"ARX\", \"KDM6B\", \"SON\", \"ATP1A3\", \"DEPDC5\", \"EN2\", \n",
    "    \"SETD1B\", \"SMARCA4\", \"WDFY3\", \"AVPR1A\", \"CDK13\", \"CHD3\", \"DLG4\", \"EBF3\", \"RORA\", \"SLC25A12\", \n",
    "    \"SLC6A8\", \"SOX5\", \"TLK2\", \"TRAPPC9\", \"ASTN2\", \"CACNA1E\", \"CNOT3\", \"CSMD1\", \"CYFIP1\", \"HECW2\", \n",
    "    \"KMT2E\", \"MACROD2\", \"MTHFR\", \"SYNE1\", \"AFF2\", \"EEF1A2\", \"MBOAT7\", \"PTCHD1\", \"SLC9A6\", \"AHI1\", \n",
    "    \"BRAF\", \"GRIK2\", \"NFIX\", \"NLGN1\", \"PHF21A\", \"CSNK2A1\", \"DSCAM\", \"GIGYF1\", \"HNRNPH2\", \"ITGB3\", \n",
    "    \"KCNQ3\", \"OPHN1\", \"RERE\", \"ARID2\", \"CC2D1A\", \"CTNNA3\", \"DHCR7\", \"LRP2\", \"LZTR1\", \"MAOA\", \n",
    "    \"MED13\", \"NIPBL\", \"PACS1\", \"RPS6KA3\", \"SEMA5A\", \"SMARCC2\", \"TANC2\", \"TNRC6B\", \"TRRAP\", \"USP9X\", \n",
    "    \"ALDH5A1\", \"ARHGEF9\", \"CHAMP1\", \"DLGAP2\", \"ITPR1\", \"KIRREL3\", \"MCPH1\", \"PRKN\", \"SMC1A\", \"UPF3B\", \n",
    "    \"AGO1\", \"GRM7\", \"HDAC4\", \"IRF2BPL\", \"JARID2\", \"KATNAL2\", \"NFIA\", \"NRXN2\", \"RAB39B\", \"SHANK1\", \n",
    "    \"SPEN\", \"SRCAP\", \"ZNF462\", \"ATP2B2\", \"CIC\", \"CNKSR2\", \"CNTN6\", \"CTNND2\", \"FBN1\", \"GLRA2\", \n",
    "    \"GRIA3\", \"KIF1A\", \"PHF8\", \"RAPGEF4\", \"SIN3A\", \"SPAST\", \"TAF1\", \"ZMYND11\", \"C12orf57\", \"CACNA2D3\", \n",
    "    \"CHRNA7\", \"CLCN4\", \"CUX2\", \"DPYD\", \"DPYSL2\", \"FBXO11\", \"HIVEP2\", \"IMMP2L\", \"KDM6A\", \"KIF5C\", \n",
    "    \"NCKAP1\", \"NLGN2\", \"NR4A2\", \"SLC6A3\", \"SLC9A9\", \"USP7\", \"WWOX\", \"ACTL6B\", \"BIRC6\", \"BRSK2\", \n",
    "    \"CELF4\", \"CNTNAP4\", \"CNTNAP5\", \"CTTNBP2\", \"DLG2\", \"GABBR2\", \"GRIA2\", \"GRIP1\", \"HOXA1\", \"LAMA1\", \n",
    "    \"RORB\", \"SETD1A\", \"SLC1A1\", \"SMC3\", \"YWHAZ\", \"ZNF804A\", \"ABCA13\", \"ARNT2\", \"ASPM\", \"CAMK2A\", \n",
    "    \"CEP290\", \"CUX1\", \"EIF4E\", \"GRIA1\", \"HCN1\", \"KCNC2\", \"KCND3\", \"KCNJ10\", \"KLF7\", \"MAP1B\", \n",
    "    \"NSD2\", \"NTRK3\", \"QRICH1\", \"RNU4-2\", \"SMAD4\", \"SRRM2\", \"TRPM3\", \"YWHAG\", \"ZNF292\", \"CACNA1B\", \n",
    "    \"CAMK2B\", \"CDK5RAP2\", \"CHD1\", \"DIP2A\", \"DPP10\", \"FAT1\", \"HERC1\", \"KANK1\", \"KAT6B\", \"MEIS2\", \n",
    "    \"NAV2\", \"PACS2\", \"PCDH15\", \"POLR3A\", \"PPP3CA\", \"PTPRD\", \"RPL10\", \"SCN3A\", \"SCN9A\", \"SNAP25\", \n",
    "    \"SPTBN1\", \"TAOK1\", \"TCF7L2\", \"ZBTB18\", \"ZFHX3\", \"ACY1\", \"ADCY5\", \"BRD4\", \"CADPS2\", \"CDC42BPB\", \n",
    "    \"CDH8\", \"CNOT1\", \"CSNK2B\", \"DHX30\", \"DLGAP1\", \"DLGAP3\", \"DMPK\", \"EP400\", \"FLNA\", \"FRMPD4\", \n",
    "    \"GABRG3\", \"HACE1\", \"HCFC1\", \"HDAC8\", \"HEPACAM\", \"HERC2\", \"HIVEP3\", \"HNRNPK\", \"INTS1\", \"KANSL1\", \n",
    "    \"KDM2B\", \"MSRA\", \"NAA10\", \"NCOR1\", \"MSNP1AS\", \"PCCA\", \"PCDH9\", \"POLR2A\", \"RFX3\", \"RIMS1\", \n",
    "    \"UBR5\", \"UNC80\", \"WDFY4\", \"ZMIZ1\", \"BSN\", \"CACNA1I\", \"CACNB2\", \"DNM1\", \"DOCK4\", \"DOCK8\", \n",
    "    \"DST\", \"FEZF2\", \"FHIT\", \"GABRG2\", \"GRB10\", \"GRM5\", \"GTF2I\", \"HECTD4\", \"HRAS\", \"KCNA2\", \n",
    "    \"KMT2B\", \"LRP1\", \"MACF1\", \"MARK1\", \"MRTFB\", \"NACC1\", \"NR3C2\", \"PAH\", \"PCCB\", \"PCDH10\", \n",
    "    \"PCLO\", \"PPM1D\", \"PRPF8\", \"PRUNE2\", \"PTPRT\", \"ST8SIA2\", \"SYT1\", \"USH2A\", \"USP24\", \"XPC\", \n",
    "    \"XPO1\", \"ACTB\", \"ARHGAP32\", \"ASMT\", \"BCL11B\", \"CACNA1F\", \"CDH10\", \"CHKB\", \"CNTN5\", \"CSMD3\", \n",
    "    \"DLX2\", \"DNAH10\", \"EPPK1\", \"GPHN\", \"GRID2\", \"H1-4\", \"HLA-DRB1\", \"ITSN1\", \"KCNC1\", \"KCNH7\", \n",
    "    \"LAMB1\", \"LRBA\", \"MAPK8IP3\", \"MYCBP2\", \"NOTCH1\", \"NRP2\", \"NTRK2\", \"OCRL\", \"PIK3CA\", \"PIK3R2\", \n",
    "    \"PLXNA3\", \"PRR12\", \"PTPRC\", \"RAC1\", \"RANBP17\", \"ROBO2\", \"SBF1\", \"SLC1A2\", \"SNX14\", \"STAG1\", \n",
    "    \"TAOK2\", \"TMLHE\", \"TRAF7\", \"UNC13A\", \"WDR26\", \"YY1\", \"ZNF536\", \"ABL2\", \"ADGRB1\", \"AGAP1\", \n",
    "    \"AKAP9\", \"AP2M1\", \"APBA2\", \"ADRB2\", \"AR\", \"ATP10A\", \"ATP1A1\", \"BCKDK\", \"BCORL1\", \"BRCA2\", \n",
    "    \"BRWD3\", \"CACNA2D1\", \"CADM1\", \"CASZ1\", \"CCDC88C\", \"CD38\", \"DLX6\", \"DMXL2\", \"DNAH17\", \"DNAH3\", \n",
    "    \"EPHB1\", \"GABRA4\", \"GIGYF2\", \"GLO1\", \"GNAS\", \"GRIK5\", \"HLA-A\", \"ICA1\", \"JMJD1C\", \"KCND2\", \n",
    "    \"KDM2A\", \"MDGA2\", \"MIB1\", \"MIR137\", \"MUC4\", \"MYO16\", \"NINL\", \"PARD3B\", \"PAX5\", \"PER1\", \n",
    "    \"PKD1\", \"POMGNT1\", \"SACS\", \"SGSH\", \"SLC12A5\", \"SLC4A10\", \"SLITRK5\", \"SND1\", \"SUPT16H\", \"SYNCRIP\", \n",
    "    \"TBC1D5\", \"TECTA\", \"TRPC5\", \"TSPAN7\", \"UNC79\", \"UPF2\", \"VDR\", \"ZMYM2\", \"ZNF559\", \"ADSL\", \n",
    "    \"AGO3\", \"ARID1A\", \"BAZ2B\", \"BTAF1\", \"AVPR1B\", \"CASKIN1\", \"CDH11\", \"CEP135\", \"COL12A1\", \"CRMP1\", \n",
    "    \"CSDE1\", \"CSMD2\", \"CTR9\", \"CUL7\", \"CYFIP2\", \"DRD3\", \"EFR3A\", \"EPHB2\", \"ERBIN\", \"ESR2\", \n",
    "    \"ESRRB\", \"FRYL\", \"GPC5\", \"HSD11B1\", \"HTR3A\", \"KCTD13\", \"KPTN\", \"MARK2\", \"MAST3\", \"MBD1\", \n",
    "    \"MEGF11\", \"MFRP\", \"MSL3\", \"MYH10\", \"MYH9\", \"NTRK1\", \"OTUD7A\", \"PAX6\", \"PBX1\", \"PLCB1\", \n",
    "    \"PPP2R1B\", \"PRKDC\", \"PRODH\", \"PTK7\", \"PXDN\", \"PTPRS\", \"SKI\", \"TAF6\", \"TLN2\", \"TNPO3\", \n",
    "    \"TRAPPC6B\", \"WNK3\", \"ZMYM3\", \"ZMYND8\", \"ABAT\", \"ABCA2\", \"ADA\", \"AGMO\", \"ANKS1B\", \"AP1S2\", \n",
    "    \"ARHGEF10\", \"ARHGEF2\", \"CAPN12\", \"CDH2\", \"CEP41\", \"CERT1\", \"CGNL1\", \"CHD4\", \"CHD9\", \"CHM\", \n",
    "    \"CLTCL1\", \"CNR1\", \"CPSF7\", \"CSNK1E\", \"DDC\", \"DIP2C\", \"EPHA1\", \"FABP5\", \"GPC6\", \"HNRNPR\", \n",
    "    \"HS3ST5\", \"HTR3C\", \"IGF1\", \"IL1R2\", \"KAT2B\", \"MED12L\", \"MEGF10\", \"METTL14\", \"MINK1\", \"MYO5A\", \n",
    "    \"MYO9B\", \"NFIB\", \"NPAS3\", \"NTNG1\", \"PAK1\", \"PC\", \"PCDHA5\", \"PDE3B\", \"PDHA1\", \"PER2\", \n",
    "    \"PHF2\", \"PLAA\", \"PPFIA3\", \"PRICKLE2\", \"PRKAR1B\", \"PRKCB\", \"PTBP2\", \"PTGS2\", \"RAB2A\", \"RPH3A\", \n",
    "    \"SCAF4\", \"SEZ6L2\", \"SGSM3\", \"SIK1\", \"SLC9A1\", \"SLITRK2\", \"STXBP5\", \"TERF2\", \"TET2\", \"TET3\", \n",
    "    \"TOP2B\", \"TOP3B\", \"TRIM23\", \"TRPM7\", \"TSHZ3\", \"TSPOAP1\", \"U2AF2\", \"ZSWIM6\", \"ACHE\", \"ACTN4\", \n",
    "    \"AGAP2\", \"AHNAK\", \"ALDH1A3\", \"ALDH1L1\", \"ANKRD17\", \"ADORA2A\", \"ARHGAP5\", \"CA6\", \"CADM2\", \"CAPRIN1\", \n",
    "    \"CCNK\", \"CDH22\", \"CDH9\", \"CDK19\", \"CHRM3\", \"CMIP\", \"CNGB3\", \"CNTNAP3\", \"CYP27A1\", \"DAGLA\", \n",
    "    \"DDHD2\", \"DDX53\", \"DRD2\", \"DUSP15\", \"EIF3F\", \"ELAVL2\", \"ELP2\", \"ELP4\", \"EXT1\", \"G3BP2\", \n",
    "    \"GABRB2\", \"GABRA2\", \"GALNT10\", \"GALNT14\", \"GNAI1\", \"GNB2\", \"GRID1\", \"GRIK3\", \"GRIK4\", \"H3-3B\", \n",
    "    \"HECTD1\", \"IKZF1\", \"IQGAP3\", \"ITGA8\", \"KIF13B\", \"KIF14\", \"LAS1L\", \"LRFN2\", \"LRFN5\", \"LRRC1\", \n",
    "    \"MAOB\", \"MBD4\", \"MSL2\", \"MTSS2\", \"MYH4\", \"NPFFR2\", \"OFD1\", \"PAPOLG\", \"PATJ\", \"PCDHA7\", \n",
    "    \"PHF14\", \"PHLPP1\", \"PITX1\", \"PLPPR4\", \"PLXNA4\", \"PNPLA7\", \"POLA2\", \"PRICKLE1\", \"PRKCA\", \"PRKD1\", \n",
    "    \"PSD3\", \"PSMD12\", \"PTPRB\", \"RAD21\", \"RALA\", \"RIMS2\", \"RLIM\", \"RSRC1\", \"SCN4A\", \"SET\", \n",
    "    \"SIN3B\", \"SLC22A9\", \"SLC23A1\", \"SNTG2\", \"SORCS3\", \"SOS 2\", \"STX1A\", \"SYN2\", \"TBC1D31\", \"TBX1\", \n",
    "    \"TSPYL2\", \"UBE3C\", \"USP27X\", \"USP45\", \"YTHDC2\", \"ZC3H4\", \"ZFYVE26\", \"ABCA7\", \"AGTR2\", \"ATP9A\", \n",
    "    \"BAIAP2L1\", \"BBS4\", \"BST1\", \"C4B\", \"CACNG2\", \"CAMK4\", \"CCDC91\", \"CDH13\", \"CLN8\", \"CLTC\", \n",
    "    \"CNTN3\", \"DCC\", \"DIXDC1\", \"DLG1\", \"DENND2B\", \"DLL1\", \"DPP4\", \"DOT1L\", \"DVL3\", \"ELAVL3\", \n",
    "    \"EPC2\", \"FBXO40\", \"GALNT13\", \"GALNT2\", \"GLI3\", \"GPR85\", \"GPX1\", \"GRK4\", \"H4C5\", \"HDLBP\", \n",
    "    \"HOMER1\", \"HTR1B\", \"ILF2\", \"INTS6\", \"KCNH1\", \"KCNH5\", \"KCNS3\", \"KDM3B\", \"KDM4B\", \"KIAA1586\", \n",
    "    \"LMTK3\", \"MAGEC3\", \"MAPK3\", \"MTF1\", \"MUC12\", \"MYO1E\", \"MYO5C\", \"NAV3\", \"NEGR1\", \"NFE2L3\", \n",
    "    \"NIPA1\", \"NLGN4Y\", \"NRCAM\", \"NSMCE3\", \"NTNG2\", \"NUAK1\", \"NUP133\", \"NXPH1\", \"P2RX5\", \"PABPC1\", \n",
    "    \"PAK2\", \"PCDHA10\", \"PCDHA11\", \"PCDHA2\", \"PCDHA4\", \"PCM1\", \"PDK2\", \"PDZD8\", \"PPFIA1\", \"PREX1\", \n",
    "    \"PRR14L\", \"PTDSS1\", \"RAB11FIP5\", \"RALGAPB\", \"RGS7\", \"SATB1\", \"SF3B1\", \"SLC45A1\", \"SLC7A5\", \n",
    "    \"SLC7A7\", \"SLCO1B3\", \"SRGAP3\", \"SRSF11\", \"SYP\", \"TBC1D23\", \"TBCK\", \"TFE3\", \"TLE3\", \"TMEM39B\", \n",
    "    \"TNRC6C\", \"TRIM8\", \"TRPC4\", \"TRPC6\", \"TRPM1\", \"TRPM6\", \"TSHZ1\", \"TUBGCP5\", \"UBR3\", \"UIMC1\", \n",
    "    \"VCP\", \"WDR37\", \"WNT1\", \"ACAP2\", \"ACE\", \"ADGRL1\", \"AGBL4\", \"AGO4\", \"AMPD1\", \"ANP32A\", \n",
    "    \"AP2S1\", \"ARF3\", \"ASAP2\", \"ASB11\", \"ASB14\", \"ATP6V0A2\", \"BCAS1\", \"BICRA\", \"CADPS\", \"CDK8\", \n",
    "    \"CELF2\", \"CELF6\", \"CHMP1A\", \"COL28A1\", \"CPT2\", \"CPZ\", \"CTNNA2\", \"CUL2\", \"CUL4B\", \"CXXC5\", \n",
    "    \"DENR\", \"DGKI\", \"DHX9\", \"DOLK\", \"DNAJC5\", \"ECPAS\", \"EIF3G\", \"EMSY\", \"ENPP1\", \"ETFB\", \"EXOC6B\", \n",
    "    \"FAN1\", \"FBXO33\", \"FGA\", \"FGFR1\", \"FRMD5\", \"GBE1\", \"GPD2\", \"GRID2IP\", \"HLA-B\", \"HNRNPD\", \n",
    "    \"HYDIN\", \"INPP1\", \"KCNN2\", \"KDM5A\", \"KDM1A\", \"LEO1\", \"LHX2\", \"LNPK\", \"LRRC4\", \"MAP4K1\", \n",
    "    \"MAPT\", \"MAST1\", \"MSR1\", \"MYO7A\", \"NAALADL2\", \"OXT\", \"P4HA2\", \"PCDH11X\", \"PCDHA1\", \"PCDHA12\", \n",
    "    \"PCDHA13\", \"PCDHA3\", \"PCDHA8\", \"PCDHA9\", \"PCDHAC1\", \"PEX7\", \"PHF12\", \"PHF3\", \"PHRF1\", \n",
    "    \"PIK3CG\", \"PLCD4\", \"PLXNB1\", \"PON1\", \"POU3F3\", \"PPP1R9B\", \"PPP2CA\", \"PPP2R5C\", \"PTPN4\", \n",
    "    \"PUF60\", \"RBBP5\", \"RBM27\", \"RNF135\", \"RPS6KA2\", \"SETDB1\", \"SH3RF1\", \"SMAP2\", \"SMG6\", \"SNX5\", \n",
    "    \"SPARCL1\", \"STK39\", \"TAF1C\", \"TAF4\", \"THBS1\", \"THRA\", \"TM9SF4\", \"TRIM32\", \"TSPAN4\", \"UBAP2L\", \n",
    "    \"UBN2\", \"UGGT1\", \"UNC5D\", \"USP15\", \"WASF1\", \"YEATS2\", \"YWHAE\", \"ZFHX4\", \"ZNF517\", \"ZNF711\", \n",
    "    \"ZWILCH\", \"ABCA10\", \"ABCE1\", \"ADCY1\", \"ADCY3\", \"ADORA3\", \"AGO2\", \"APBB1\", \"ARHGAP11B\", \n",
    "    \"AZGP1\", \"BACE1\", \"BRINP3\", \"BTRC\", \"CAMK2D\", \"CAT\", \"CCSER1\", \"CD276\", \"CDON\", \"CLASP1\", \n",
    "    \"CMPK2\", \"CPEB4\", \"CX3CR1\", \"CYLC2\", \"DDX23\", \"CYP11B1\", \"DLX3\", \"DMWD\", \"DOCK1\", \"DPP3\", \n",
    "    \"DRD1\", \"EIF4G1\", \"EIF5A\", \"FAM98C\", \"FBRSL1\", \"FCRL6\", \"FGF13\", \"GALNT8\", \"H4C11\", \"HLA-G\", \n",
    "    \"HNRNPF\", \"HNRNPUL2\", \"KATNAL1\", \"KDM3A\", \"KDM4C\", \"KHDRBS2\", \"KIAA0232\", \"KLHL20\", \"KRT26\", \n",
    "    \"LARP1\", \"LDB1\", \"LDLR\", \"LEMD3\", \"LILRB2\", \"LMX1B\", \"MAP1A\", \"MAPK8IP1\", \"MBD3\", \"MCM4\", \n",
    "    \"MCM6\", \"MED23\", \"LZTS2\", \"MKX\", \"MLANA\", \"MYLK\", \"NCKAP5\", \"NCOA1\", \"NEDD4\", \"NEO1\", \n",
    "    \"NR1D1\", \"NMT1\", \"NPAS2\", \"NUDCD2\", \"NUP155\", \"OR1C1\", \"OR2T10\", \"OTX1\", \"PAFAH1B2\", \"PCDHA6\", \n",
    "    \"PCDHAC2\", \"PDCD1\", \"PHF7\", \"PLN\", \"POMT1\", \"PRKD2\", \"PTCHD1-AS\", \"PYHIN1\", \"RAP1A\", \n",
    "    \"RASSF5\", \"RFX7\", \"RICTOR\", \"RIMS3\", \"RNF25\", \"RUNX1T1\", \"SAE1\", \"SAMD11\", \"SASH1\", \"SCAF1\", \n",
    "    \"SCFD2\", \"SCGN\", \"RPS10P2-AS1\", \"SDC2\", \"SENP6\", \"SERPINE1\", \"SH3RF3\", \"SLC24A2\", \"SLC35G1\", \n",
    "    \"SLFN5\", \"SOX6\", \"SPP2\", \"SPRY2\", \"SRSF1\", \"SSR4\", \"ST7\", \"STYK1\", \"SYT17\", \"TCEAL1\", \n",
    "    \"TEK\", \"TGM1\", \"TM4SF19\", \"TM4SF20\", \"TNS2\", \"TRAP1\", \"TRAPPC2L\", \"TSPAN17\", \"UBR1\", \n",
    "    \"TDO2\", \"VAMP2\", \"VASH1\", \"VIL1\", \"VSIG4\", \"ZBTB21\", \"ZC3H11A\", \"ZFX\", \"ZNF532\", \"ZNF548\", \n",
    "    \"ZNF626\", \"ZNF774\", \"ZNF827\", \"ADSS2\", \"AMT\", \"APH1A\", \"ARHGAP30\", \"ATP2B1\", \"ATXN2\", \n",
    "    \"C15orf62\", \"CACNB1\", \"CAMTA2\", \"CBX1\", \"CCT4\", \"CAPZA2\", \"CBX4\", \"CDK16\", \"CECR2\", \"CHRNB3\", \n",
    "    \"CHST2\", \"CORO1A\", \"CSNK1G1\", \"CTNND1\", \"DAPP1\", \"CTPS1\", \"DIPK2A\", \"DNER\", \"DPYSL3\", \n",
    "    \"EGR3\", \"EXOC3\", \"EXOC6\", \"FABP4\", \"FBXL13\", \"FRG1\", \"FRK\", \"GABRA3\", \"GDA\", \"GGNBP2\", \n",
    "    \"GLIS1\", \"GNB1L\", \"GPC4\", \"GPR37\", \"GUCY1A2\", \"H2BC11\", \"H4C3\", \"GSTM1\", \"IL1RAPL2\", \"IRX5\", \n",
    "    \"KCNJ15\", \"KCNK7\", \"KDM1B\", \"KRR1\", \"LIN7B\", \"LRRC4C\", \"MBD6\", \"MEMO1\", \"MYOCD\", \"NASP\", \n",
    "    \"NIPA2\", \"NOVA2\", \"NPTN\", \"NDUFA5\", \"OR52M1\", \"PDE1C\", \"PLAUR\", \"POT1\", \"PPP5C\", \"PRPF19\", \n",
    "    \"PRPF39\", \"PSMC5\", \"PSMD11\", \"PSMD6\", \"RAB11FIP4\", \"RAD21L1\", \"REEP3\", \"RFX4\", \"RHEB\", \n",
    "    \"RHOXF1\", \"RIT2\", \"SCP2\", \"SENP1\", \"SETDB2\", \"SHOX\", \"SLC25A39\", \"SLC27A4\", \"SLC38A10\", \n",
    "    \"SLC7A3\", \"SMURF1\", \"SNCAIP\", \"SOD1\", \"SRPRA\", \"SSRP1\", \"SYCE1\", \"SYNJ1\", \"TBCEL\", \"TBCB\", \n",
    "    \"TMEM134\", \"TTI2\", \"USP30\", \"VEZF1\", \"VPS54\", \"VWA7\", \"WDR5\", \"ZBTB16\", \"ZNF385B\", \"ZNF644\", \n",
    "    \"ZNF713\", \"ADK\", \"AGAP5\", \"ALG6\", \"ANXA1\", \"BICDL1\", \"CARD11\", \"CCIN\", \"CCNG1\", \"CIB2\", \n",
    "    \"CLIP2\", \"DYDC1\", \"ELOVL2\", \"ENOX2\", \"EXOC5\", \"FAM47A\", \"FGF14\", \"FRRS1L\", \"FXN\", \"GATM\", \n",
    "    \"GFAP\", \"HLA-DPB1\", \"HTR2C\", \"HNRNPL\", \"KCNA3\", \"KIZ\", \"KNG1\", \"MAP4K4\", \"METTL26\", \n",
    "    \"MSANTD2\", \"MSX2\", \"NKX2-2\", \"NXF1\", \"PEBP4\", \"PLEKHA8\", \"PPP1R1B\", \"PRR25\", \"RAB43\", \n",
    "    \"RNF38\", \"RNU2-2\", \"RNU5B-1\", \"SLC29A4\", \"SYAP1\", \"TBX22\", \"TPO\", \"TRIM33\", \"UBE2H\", \n",
    "    \"USP9Y\", \"WWP1\", \"XRCC6\", \"ZBTB47\", \"ZBTB7A\", \"ZNF18\", \"ASB9\", \"CD99L2\", \"DYDC2\", \"ERG\", \n",
    "    \"ERMN\", \"GAS2\", \"HMGN1\", \"KLF16\", \"LEP\", \"MAPT-AS1\", \"NCAPH2\", \"MNT\", \"OR2M4\", \"PHB1\", \n",
    "    \"PJA1\", \"RP11-1407O15.2\", \"SLC22A15\", \"SLC25A27\", \"SLC35B1\", \"SYBU\", \"TBL1X\", \"TERB2\", \n",
    "    \"TFB2M\"\n",
    "}\n",
    "\n",
    "# ========================================================================\n",
    "# 3. BUILD INTERVALS \n",
    "# ========================================================================\n",
    "print(\"Downloading Gene Map to filter intervals...\")\n",
    "gencode_url = \"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/gencode.v44.basic.annotation.gtf.gz\"\n",
    "gene_map_df = pd.read_csv(gencode_url, sep='\\t', comment='#', header=None, names=['chrom', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes'])\n",
    "\n",
    "def extract_gene_name(attr):\n",
    "    if 'gene_name \"' in attr:\n",
    "        return attr.split('gene_name \"')[1].split('\"')[0]\n",
    "    return None\n",
    "\n",
    "gene_map_df = gene_map_df[gene_map_df['feature'] == 'gene']\n",
    "gene_map_df['gene_name'] = gene_map_df['attributes'].apply(extract_gene_name)\n",
    "target_intervals_df = gene_map_df[gene_map_df['gene_name'].isin(asd_gene_target_list)]\n",
    "\n",
    "intervals = []\n",
    "for _, row in target_intervals_df.iterrows():\n",
    "    chrom = str(row['chrom']).replace(\"chr\", \"\")\n",
    "    final_chrom = f\"chr{chrom}\"\n",
    "    intervals.append(hl.parse_locus_interval(f\"[{final_chrom}:{row['start']}-{row['end']}]\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading MT with Interval Filter...\n",
      "✓ Speed Hack Applied.\n"
     ]
    }
   ],
   "source": [
    "# ========================================================================\n",
    "# 4. LOAD & FILTER\n",
    "# ========================================================================\n",
    "print(\"Reading MT with Interval Filter...\")\n",
    "mt = hl.read_matrix_table(wgs_path)\n",
    "\n",
    "if intervals:\n",
    "    # This is the speed hack\n",
    "    mt = hl.filter_intervals(mt, intervals)\n",
    "    print(\"✓ Speed Hack Applied.\")\n",
    "else:\n",
    "    print(\"Warning: No intervals found.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# ========================================================================\n",
    "# 5. FILTER COHORTS (10 PEOPLE)\n",
    "# ========================================================================\n",
    "if 'dataset_1_person_df' in locals():\n",
    "    asd_ids_pool = dataset_63330128_person_df['person_id'].astype(str).tolist()\n",
    "else:\n",
    "    asd_ids_pool = [] \n",
    "\n",
    "if 'dataset_2_person_df' in locals():\n",
    "    control_ids_pool = dataset_84568391_person_df['person_id'].astype(str).tolist()\n",
    "else:\n",
    "    control_ids_pool = []\n",
    "\n",
    "if asd_ids_pool and control_ids_pool:\n",
    "    n_select = 5\n",
    "    selected_asd_ids = random.sample(asd_ids_pool, min(len(asd_ids_pool), n_select))\n",
    "    selected_control_ids = random.sample(control_ids_pool, min(len(control_ids_pool), n_select))\n",
    "    all_sample_ids = selected_asd_ids + selected_control_ids\n",
    "    \n",
    "    print(f\"Filtering to 10 Individuals: {all_sample_ids}\")\n",
    "    mt = mt.filter_cols(hl.literal(all_sample_ids).contains(mt.s))\n",
    "else:\n",
    "    print(\"Warning: Cohort IDs not found. Skipping sample filter.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exporting Variants...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 0:==================================================>(13443 + 1) / 13444]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✓ Variants Exported to gs://fc-secure-b8cbf1c7-c51b-44be-b4f7-4f88082bda65/data/aou_sfari_variants_fast.csv\n",
      "Exporting Genotypes...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 1:==================================================>(13443 + 1) / 13444]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✓ Genotypes Exported to gs://fc-secure-b8cbf1c7-c51b-44be-b4f7-4f88082bda65/data/aou_sfari_genotypes_long.csv\n",
      "\n",
      "SUCCESS! You are done with extraction.\n"
     ]
    }
   ],
   "source": [
    "# ========================================================================\n",
    "# 6. EXPORT\n",
    "# ========================================================================\n",
    "print(\"Exporting Variants...\")\n",
    "rows_ht = mt.rows()\n",
    "\n",
    "variants_table = rows_ht.select(\n",
    "    Chrom = rows_ht.locus.contig,\n",
    "    Pos = rows_ht.locus.position,\n",
    "    Ref = rows_ht.alleles[0],\n",
    "    Alt = rows_ht.alleles[1],\n",
    "    Gene = hl.literal(\"Lookup_Required\")\n",
    ")\n",
    "\n",
    "variants_filename = f\"{bucket}/data/aou_sfari_variants_fast.csv\"\n",
    "variants_table.export(variants_filename, delimiter=',')\n",
    "print(f\"✓ Variants Exported to {variants_filename}\")\n",
    "\n",
    "print(\"Exporting Genotypes...\")\n",
    "entries_ht = mt.entries()\n",
    "\n",
    "entries_table = entries_ht.select(\n",
    "    Var_ID = hl.str(entries_ht.locus) + \"_\" + entries_ht.alleles[0] + \">\" + entries_ht.alleles[1],\n",
    "    Sample_ID = entries_ht.s,\n",
    "    GT_num = hl.if_else(hl.is_defined(entries_ht.GT), entries_ht.GT.n_alt_alleles(), 0)\n",
    ")\n",
    "\n",
    "genotypes_filename = f\"{bucket}/data/aou_sfari_genotypes_long.csv\"\n",
    "entries_table.export(genotypes_filename, delimiter=',')\n",
    "print(f\"✓ Genotypes Exported to {genotypes_filename}\")\n",
    "\n",
    "print(\"\\nSUCCESS! You are done with extraction.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil ls $WORKSPACE_BUCKET/data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# patient-level per person gene module score and heatmap \n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import requests\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from tqdm import tqdm\n",
    "\n",
    "# 1. SETUP & LOAD DATA\n",
    "# --------------------\n",
    "bucket = os.getenv(\"WORKSPACE_BUCKET\")\n",
    "print(\"Loading Patient Data...\")\n",
    "\n",
    "# Ensure we have the variant & genotype data\n",
    "if not os.path.exists(\"aou_sfari_variants_fast.csv\"):\n",
    "    os.system(f\"gsutil cp {bucket}/data/aou_sfari_variants_fast.csv .\")\n",
    "    os.system(f\"gsutil cp {bucket}/data/aou_sfari_genotypes_long.csv .\")\n",
    "    os.system(f\"gsutil cp {bucket}/results/final_patient_results.csv .\") # Scores from Phase 2\n",
    "\n",
    "variants_df = pd.read_csv(\"aou_sfari_variants_fast.csv\")\n",
    "genotypes_df = pd.read_csv(\"aou_sfari_genotypes_long.csv\")\n",
    "\n",
    "# We need the calculated risk scores from Phase 2 (re-calculating merge if needed)\n",
    "# Ideally, Phase 2 added 'risk_score' to variants_df. \n",
    "# If that dataframe is lost in memory, let's assume we need to reload or re-merge.\n",
    "# For safety, let's assume you ran Phase 2 and variants_df HAS 'risk_score'.\n",
    "# If not, we re-load the saved result if available, or just re-calculate quickly.\n",
    "# (Here I will assume variants_df has the columns: Chrom, Pos, risk_score. \n",
    "#  If 'risk_score' is missing, please re-run Phase 2 or let me know).\n",
    "\n",
    "# 2. RE-ANNOTATE GENES (Map Chr:Pos -> Gene Symbol)\n",
    "# --------------------\n",
    "print(\"Mapping Variants to Gene Symbols...\")\n",
    "\n",
    "# Download GENCODE again (fastest way to map)\n",
    "gencode_url = \"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/gencode.v44.basic.annotation.gtf.gz\"\n",
    "gene_map_df = pd.read_csv(gencode_url, sep='\\t', comment='#', header=None, names=['chrom', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes'])\n",
    "\n",
    "def extract_gene_name(attr):\n",
    "    if 'gene_name \"' in attr:\n",
    "        return attr.split('gene_name \"')[1].split('\"')[0]\n",
    "    return None\n",
    "\n",
    "# Filter to Genes only\n",
    "gene_map = gene_map_df[gene_map_df['feature'] == 'gene'].copy()\n",
    "gene_map['gene_name'] = gene_map['attributes'].apply(extract_gene_name)\n",
    "gene_map = gene_map[['chrom', 'start', 'end', 'gene_name']]\n",
    "\n",
    "# Create an Interval Index for super-fast lookup\n",
    "# We use pandas IntervalIndex to find which gene range a variant falls into\n",
    "# Note: This ignores overlapping genes for simplicity (taking the first match)\n",
    "import pandas as pd\n",
    "\n",
    "# Clean chrom names\n",
    "gene_map['chrom'] = gene_map['chrom'].astype(str)\n",
    "# Handle prefix matching\n",
    "if not str(variants_df['Chrom'].iloc[0]).startswith('chr'):\n",
    "    gene_map['chrom'] = gene_map['chrom'].str.replace('chr', '')\n",
    "\n",
    "# Function to map variant to gene\n",
    "# This simple loop is robust for small datasets (~1000 variants). \n",
    "# For millions, we'd use IntervalIndex.\n",
    "def map_variant_to_gene(chrom, pos):\n",
    "    # Find genes on this chromosome where pos is between start and end\n",
    "    # We filter gene_map down to just the relevant chrom first\n",
    "    candidates = gene_map[\n",
    "        (gene_map['chrom'] == str(chrom)) & \n",
    "        (gene_map['start'] <= pos) & \n",
    "        (gene_map['end'] >= pos)\n",
    "    ]\n",
    "    if not candidates.empty:\n",
    "        return candidates.iloc[0]['gene_name']\n",
    "    return \"Unknown\"\n",
    "\n",
    "tqdm.pandas(desc=\"Annotating Genes\")\n",
    "variants_df['Gene_Symbol'] = variants_df.apply(lambda x: map_variant_to_gene(x['Chrom'], x['Pos']), axis=1)\n",
    "\n",
    "print(f\"Annotated {variants_df['Gene_Symbol'].nunique()} unique genes.\")\n",
    "\n",
    "# 3. LOAD PATHWAY DATABASE (Enrichr GMT)\n",
    "# --------------------\n",
    "print(\"Downloading Gene Ontology (GO) Biological Process...\")\n",
    "# Standard GO Biological Process 2023 from Enrichr\n",
    "gmt_url = \"https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName=GO_Biological_Process_2023\"\n",
    "response = requests.get(gmt_url)\n",
    "\n",
    "# Parse GMT: PathwayName \\t Description \\t Gene1 \\t Gene2 ...\n",
    "pathway_dict = {}\n",
    "for line in response.text.splitlines():\n",
    "    parts = line.split('\\t')\n",
    "    pathway_name = parts[0]\n",
    "    genes = set(parts[2:]) # Skip description\n",
    "    pathway_dict[pathway_name] = genes\n",
    "\n",
    "print(f\"Loaded {len(pathway_dict)} pathways.\")\n",
    "\n",
    "# Filter to only relevant pathways (ASD-related keywords)\n",
    "# This reduces noise and processing time\n",
    "relevant_keywords = ['synap', 'neuron', 'axon', 'dendri', 'ion channel', 'chromatin', 'behavior', 'learning']\n",
    "selected_pathways = {k: v for k, v in pathway_dict.items() if any(w in k.lower() for w in relevant_keywords)}\n",
    "print(f\"Filtered to {len(selected_pathways)} neuro-relevant pathways.\")\n",
    "\n",
    "# 4. CALCULATE SCORES PER PATHWAY\n",
    "# --------------------\n",
    "print(\"Calculating Scores per Patient per Pathway...\")\n",
    "\n",
    "# Prepare the \"Patient-Gene\" Table first\n",
    "# We aggregate risk at the gene level for each patient\n",
    "variants_df['Var_ID'] = variants_df['Chrom'].astype(str) + \":\" + variants_df['Pos'].astype(str) + \"_\" + variants_df['Ref'] + \">\" + variants_df['Alt']\n",
    "merged = genotypes_df.merge(variants_df[['Var_ID', 'Gene_Symbol', 'risk_score']], on='Var_ID')\n",
    "merged = merged[merged['GT_num'] > 0] # Only actual mutations\n",
    "merged['burden'] = merged['GT_num'] * merged['risk_score']\n",
    "\n",
    "# Sum scores per Gene per Patient\n",
    "patient_gene_scores = merged.groupby(['Sample_ID', 'Gene_Symbol'])['burden'].sum().reset_index()\n",
    "\n",
    "# Now aggregate into Pathways\n",
    "patient_pathway_scores = []\n",
    "\n",
    "for patient_id in patient_gene_scores['Sample_ID'].unique():\n",
    "    pat_data = patient_gene_scores[patient_gene_scores['Sample_ID'] == patient_id]\n",
    "    \n",
    "    # Create a set of this patient's mutated genes for fast lookup\n",
    "    pat_genes = dict(zip(pat_data['Gene_Symbol'], pat_data['burden']))\n",
    "    \n",
    "    for path_name, path_genes in selected_pathways.items():\n",
    "        # Score = Sum of burdens of genes in this pathway\n",
    "        score = 0\n",
    "        overlap_count = 0\n",
    "        for gene in path_genes:\n",
    "            if gene in pat_genes:\n",
    "                score += pat_genes[gene]\n",
    "                overlap_count += 1\n",
    "        \n",
    "        if score > 0: # Only store non-zero scores\n",
    "            patient_pathway_scores.append({\n",
    "                'Person_ID': patient_id,\n",
    "                'Pathway': path_name.split(' (GO')[0], # Clean name\n",
    "                'Pathway_Score': score,\n",
    "                'Gene_Count': overlap_count\n",
    "            })\n",
    "\n",
    "pathway_df = pd.DataFrame(patient_pathway_scores)\n",
    "\n",
    "# 5. VISUALIZATION (Heatmap)\n",
    "# --------------------\n",
    "if not pathway_df.empty:\n",
    "    # Pivot for Heatmap: Rows=Pathway, Cols=Patient\n",
    "    matrix = pathway_df.pivot_table(index='Pathway', columns='Person_ID', values='Pathway_Score', fill_value=0)\n",
    "    \n",
    "    # Filter for top active pathways (e.g., top 20 by max score) to keep plot readable\n",
    "    top_pathways = matrix.max(axis=1).sort_values(ascending=False).head(20).index\n",
    "    matrix_top = matrix.loc[top_pathways]\n",
    "\n",
    "    plt.figure(figsize=(12, 10))\n",
    "    sns.heatmap(matrix_top, cmap=\"Reds\", annot=True, fmt=\".1f\", linewidths=.5)\n",
    "    plt.title(\"Pathogenicity Score by Biological Pathway per Patient\")\n",
    "    plt.xlabel(\"Patient ID\")\n",
    "    plt.ylabel(\"GO Biological Process\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"patient_pathway_heatmap.png\")\n",
    "    plt.show()\n",
    "    \n",
    "    # Save Results\n",
    "    os.system(f\"gsutil cp patient_pathway_heatmap.png {bucket}/results/\")\n",
    "    pathway_df.to_csv(\"final_patient_pathway_scores.csv\", index=False)\n",
    "    os.system(f\"gsutil cp final_patient_pathway_scores.csv {bucket}/results/\")\n",
    "    print(f\"\\n✓ Saved Pathway Scores to {bucket}/results/final_patient_pathway_scores.csv\")\n",
    "    \n",
    "    # \n",
    "else:\n",
    "    print(\"No non-zero pathway scores found (Did the patients have variants in the gene list?)\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
